File: DotNetflixServer/DotNetflix.Abstractions/Result.cs
namespace DotNetflix.Abstractions;

public class Result<TSuccess, TFailure>
{
    private readonly TSuccess? _success;
    private readonly TFailure? _failure;
    private readonly bool _isSuccess;

    public Result(TSuccess success)
    {
        _success = success;
        _isSuccess = true;
    }

    public Result(TFailure failure)
    {
        _failure = failure;
        _isSuccess = false;
    }

    public TResult Match<TResult>(Func<TSuccess, TResult> success, Func<TFailure, TResult> failure)
    {
        return _isSuccess
            ? success(_success!)
            : failure(_failure!);
    }

    public static implicit operator Result<TSuccess, TFailure>(TSuccess success)
    {
        return new Result<TSuccess, TFailure>(success);
    }

    public static implicit operator Result<TSuccess, TFailure>(TFailure failure)
    {
        return new Result<TSuccess, TFailure>(failure);
    }
}

File: mobile/lib/utils/result.dart
class Result<TSuccess, TFailure> {
  
  final TSuccess? _success;
  final TFailure? _error;
  final bool _isSuccess;

  const Result.fromSuccess(this._success) : _error = null, _isSuccess = true;
  const Result.fromFailure(this._error) : _success = null, _isSuccess = false;

  TResult match<TResult>(TResult Function(TSuccess) onSuccess, TResult Function(TFailure) onError) {
    return _isSuccess 
      ? onSuccess(_success as TSuccess) 
      : onError(_error as TFailure);
  }
}